### PR TITLE
Proposal doc [WIP]

### PR DESCRIPTION
**Pull request by ggaurav10**
Adding proposal doc

### CODE REVIEW

**Review by vlerenc — COMMENTED** [review]
HVPA is primarily/only beneficial to the API server, right? Since we have some demand for control plane autoscaling, do we plan any intermediate steps (maybe the wrong place to discuss in this PR)? E.g. VPA for all components but the API server and HPA on CPU for the API servers and VPA (initial or auto) for their memory? I am trying to find a good intermediate plan before we dive fully into HVPA.

File: Proposal.md [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+* Works even if HPA and VPA act on different metrices
+* Disruptions because of rolling updates can be controlled as rolling updates will be applied less frequently because of higher threshold
+* Provides more control to users for autoscaling apps
+* Relies on users' understanding of application, and its scaling
+#### Cons:
+* Relies on users' understanding of application, and its scaling
+* Will need both VPA and HPA to be deployed in a recommendation-only mode.
+
+### Weight based scaling
+In this mode, deployment's scaling is divided into 3 stages depending on number of replicas, and a user can provide an interval `x1` to `x2` and 3 `vpaWeight`s between and including `0` and `1`. Stage 1 will be from HPA's `minReplicas` to `x1`. Stage 2 will be from `x1` to `x2`. Stage 3 will be from `x2` to HPA's `maxReplicas`. HVPA controller will consider VPA's recommendations according to the weights provided, and scale the deployment accordingly. The weights given to the HPA's recommendation will be `1 - vpaWeight`.
```
[the remark this answers is not in the document]
Comment (amshuman-kr) [reply Proposal.md]: 👍 Even the API would be more elegant :-)

**Comment by amshuman-kr** [comment]
> HPA on CPU for the API servers and VPA (initial or auto) for their memory?

The way VPA updates are done today (using admission controller) makes me nervous to mix them for `kube-apiserver` as they are. But if we use them for different metrics, I guess this could be mitigated.

However, if the CPU base request is low, VPA might be better for CPU as well. Maybe there is a tradeoff here by raising the CPU base request.

File: Proposal.md [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+
+According to VPA, new `requests` will be `currentRequest + (targetRequest - currentRequest) * vpaWeight`
```
Comment (georgekuruvillak) [inline Proposal.md]: Should we have one more parameter alpha[0,1]
According to VPA, new requests will be currentRequest + (targetRequest - currentRequest) * vpaWeight * alpha

According to HPA, new number of replicas will be ceil of currentReplicas + (desiredReplicas - currentReplicas) * (1 - hpaWeight) * (1 - alpha)

Then we have [hpa/vpa]Weight for how much confidence we should provide for the corresponding reccomendations from hpa/vpa and use alpha for toggling between hpa and vpa decisions.

File: Proposal.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (ggaurav10) [reply Proposal.md]: Addressed here:
https://github.com/ggaurav10/hvpa-controller/pull/1/files#diff-d832191f0dcd62ab50efa1b3fe7bab96R288

User can provide multiple such intervals in this approach.

**Comment by amshuman-kr** [comment]
> do we plan any intermediate steps (maybe the wrong place to discuss in this PR)? E.g. VPA for all components but the API server and HPA on CPU for the API servers and VPA (initial or auto) for their memory? I am trying to find a good intermediate plan before we dive fully into HVPA.

@vlerenc I think the effort for implementing HVPA is less as compared to the effort in coming up with the right values for the different weights to ensure that kube-apiserver scales well. It might take more effort to home in on the right values to get scale down to work properly.

So, another alternative intermediate plan could be that we implement the weight-based approach but only for scaling up. That too, without the changes to HPA so that we scale out fully first and then scale up (i.e. weights [0, 1]). For the time being, we need not implement scale down vertically at all and let only HPA do the scale down. This way we can address both memory and CPU metrics in HPA as well as VPA. This can be implemented without touching HPA or kube-controller-manager. 

To be sure, this might over-provision for some clusters that see a rare large burst but HPA scaling down might mitigate it somewhat and the clusters ((both small and large) that do not see a large burst will reach optimal scale and remain there.

We can introduce scaling down (and simultaneous horizontal and vertical scaling) in due course. WDYT?

**Comment by vlerenc** [comment]
@amshuman-kr What was our experience from using HPA for CPU and VPA for memory?

In general, what have we learned about VPA? My understanding is, that we have not yet sufficient knowledge/experience. We don't use it yet for any control plane component. We have not yet optimised the shoot cluster components either. And yet we want to build a new component already.

Don't get me wrong. I am just worried, that we dive into this component and make no progress on the scale-up/out of our clusters, but it's critical to come up with solutions there. People use our clusters and hit scale-up/out issues that we compensate with more and more CPU and memory.

**Comment by amshuman-kr** [comment]
> What was our experience from using HPA for CPU and VPA for memory?
Not good. Only CPU scaling horizontally will need a raise the base CPU request. VPA with `Initial` update policy for memory is still ok.

As part of [#893])(https://github.com/gardener/gardener/issues/893), we are planning to introduce VPA with `Off` update policy for kube-apiserver as a part of the vpa-exporter change to gardener. So, hopefully we will have a good handle on the VPA recommendations for kube-apiserver along with other control-plane components even before we have the intermediate approach ready.

**Comment by vlerenc** [comment]
I have difficulties understanding the use cases for the staged approach, especially because of the (mentioned) problem that you need to understand the application that’s to be scaled well enough, but we don’t and can’t foresee the usage (e.g. of an API server). Which application would be scaled only horizontally for some time before you scale vertically and where would you decide is this break-off point? Min (horz.) replica makes sense, of course (e.g. 2 or 3 if you want HA). So, how would you find your stage demarcations? Maybe that’s just a theoretical argument. In my eyes, any solution that requires this kind of tuning is problematic, because we can’t provide the ideal configuration and performance measurements won’t help either.

A few questions:
- I am wondering how continuous the recommendations are written and is there really often a time window where we see both in parallel?
- What is HPA in recommendation mode anyhow – how is that done?
- In that context as well, `scaleUp/Down/Out/InDelays` looks helpful (defers until more recommendations are available). By waiting, we might see more recommendations. Hopefully, we don’t wait too long, though (crash), so I am confused. Normally there is no `scaleUp/OutDelay`, because that could cause a sudden crash? Shouldn’t we only therefore have `scaleDown/InDelay` to avoid frequent oscillation?
- `minCpuChange` and `minMemChange` is only for vertical scaling, right (described in doc)? Can’t we then name it accordingly or put it somewhere into the vpaTemplate?
- The example on `minCpuChange` and `minMemChange` only mentions scale up, but not down. And could I avoid oscilliation around one narrow band of values at all (hysteresis)?

How would we envision an API server configuration (sorry, if I always return to this use case, but I learn from examples, before I can hope to find the proper abstraction – ideally different examples that help me understand what the abstraction needs to offer)?

```
apiVersion: autoscaling.k8s.io/v1alpha1
kind: Hvpa
metadata:
  name: hvpa-api-server
spec:
  scaleOutDelay:  ~
  scaleUpDelay:   ~
  scaleInDelay:   "15m"
  scaleDownDelay: "60m"
  minCpuChange:
    value: 50m
    percentage: 70
  minMemChange:
    value: 200Mi
    percentage: 80
  hpaTemplate:
    minReplicas: 2
    maxReplicas: 10
    metrics:
    - type: Resource
      resource:
        name: cpu
        target:
          type: Utilization
          averageUtilization: 50
    - type: Resource
      resource:
        name: memory
        target:
          type: Utilization
          averageUtilization: 50
  vpaTemplate:
    resourcePolicy:
      containerPolicies:
      - containerName: api-server
	minAllowed:
          cpu: 50m
          memory: 200Mi
        maxAllowed:
          cpu: 12000m
          memory: 48000Mi
  weightBasedScalingIntervals:
    - vpaWeight: 0.4
      startReplicaCount: 2
      lastReplicaCount: 10
  targetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: api-server
```

Can we develop that further together? What is even the ideal curve (out/up and in/down)? It’s not scaling horizontally until we hit 10 replicas. Maybe 2->3, but before we go to 4, we’d rather double vertically and scale back to 2. Can we find first a picture that illustrates what we believe would make sense here and then come up with the configuration? If we continue this pattern, wouldn’t that mean, that we always scale 2->3 and then double while going back to 2? How do the recommendations from VPA and HPA differ exactly and would that influence the bahavior? But do we really want that/why would they differ in a truely meaningful way? I mean, why would the HPA recommendation hold true only for horizontal scaling and VPA for vertical? Both should "see" the same need, i.e. more or less CPU/memory and it's our decision what we do then.

**Comment by amshuman-kr** [comment]
@vlerenc To address the point about ideal curve for `kube-apiserver`, based on the performance results observed so far, for memory and CPU based scaling, it looks best to scale horizontally between 1-2 replicas and then (mostly) vertically until we reach a reasonable (based on node capacity) maxAllowed for VPA and then scale horizontally after that (may be till 5 or 6, but this could increase if we activate VPA for etcd).

But in the intermediate step, we are not tackling the controling of HPA scaling. So, for the intermediate step, we will scale horizontally from 1-3 (or 1-4) and then scale vertically. Clearly, this is not ideal but to do the more optimal way (as describe above) would require enhancing kube-control-manager which might be a longer term task. So, it was skipped for the time being. We should pursue that actively even if the intermediate step gives good benefit.

**Comment by amshuman-kr** [comment]
The possibility of crashes because of delays due to stabilisation windows could be mitigated by using safe percentage utilisation values in the scaling metrics and safe limits too.

**Comment by amshuman-kr** [comment]
The delays/stabilization windows for scaling up and down can be specified separately. Typically, the scale down windows would be much longer (orders of magnitude?). This would provide some form of hysteresis to safeguard against oscillation. Again, not ideal but may be we can improve this based on experience.

**Comment by amshuman-kr** [comment]
Regarding apiserver configuration, I am not sure if hvpa (or hpa or vpa for that matter) is the right tool to tacke that or not. I think we might need a separate component.

**Comment by amshuman-kr** [comment]
Another fly in the ointment is that HPA can scale based on custom metrics and VPA can't. So, if the recommendation is coming from HPA based on some custom metric like network IO, then it might be better to honour HPAs recommendation even if the weightage is not favourable to HPA. But it looks like we might need some enhancements in HPA status to detect such cases before we can take the right action.

**Comment by vlerenc** [comment]
> Regarding apiserver configuration, I am not sure if hvpa (or hpa or vpa for that matter) is the right tool to tacke that or not. I think we might need a separate component.

@amshuman-kr What else if not the API server is the target of HVPA? There is no other horizontally scaling component in the control plane and we ought to start learning somewhere/improve the API server scale-out anyway. You also implicitly agreed when I asked this question at the beginning of this thread: "...to ensure that kube-apiserver scales well".

If we have no concrete component to target with HVPA, we should not do it. If we have, and I thought it's the API server, then why not plot the desired/optimal scale-out/up curve and make HVPA handle it accordingly?

**Comment by amshuman-kr** [comment]
> @amshuman-kr What else if not the API server is the target of HVPA?

**Comment by amshuman-kr** [comment]
@ggaurav10 Is this PR now stale after [gardener/hvpa-controller#74](https://github.com/gardener/hvpa-controller/pull/74)?

**Pull request closed without merging**